### PR TITLE
fix: internal review findings

### PR DESCRIPTION
**Pull request by 0xChin**
work in progress allies

### ✨ PR Description
Purpose: Fix internal review findings in PolicyEngineStaking contract by refining ownership model, beneficiary semantics, and type safety with comprehensive validation improvements.

Main changes:
- Replaced immutable owner with mutable ownership supporting transfer, added ISemver interface, and downgraded uint256 to uint128 for amounts
- Renamed "link/unlink" terminology to "beneficiary set/remove" with auto-reset on allowlist revocation and same-beneficiary validation
- Added zero-address validation in constructor, new custom errors, and comprehensive test coverage for ownership transfer and allowlist edge cases

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using.
💡 **Tip:** You can customize your AI Description using **Guidelines** [Learn how](https://docs.gitstream.cm/automation-actions/#describe-changes)</sub>

### CODE REVIEW

**Review by ashitakah — DISMISSED** [review]

**Review by ashitakah — DISMISSED** [review]
LGTM!

File: packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol [hunk 1]
```diff
@@ -149,102 +180,111 @@ contract PolicyEngineStaking {
...
-            emit Linked(msg.sender, _beneficiary);
-        } else if (currentLink != _beneficiary) {
-            // Re-linking: move existing stake from old beneficiary to new
-            _decreasePeData(currentLink, uint128(stakedData.stakedAmount));
-            emit Unlinked(msg.sender, currentLink);
-
-            stakedData.linkedTo = _beneficiary;
-            _increasePeData(_beneficiary, uint128(stakedData.stakedAmount));
-            emit Linked(msg.sender, _beneficiary);
+        StakedData storage stakingData = stakingData[msg.sender];
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol]: i thought we purposely changed this in the past to not be confusing with the mapping, we should keep the `stakedData` or rename

File: packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol [hunk 2]
```diff
@@ -149,102 +180,111 @@ contract PolicyEngineStaking {
...
-    /// @notice Allows or denies a staker to attribute ordering power to the caller.
+    /// @notice Sets whether a staker can set the caller as beneficiary.
     /// @param _staker The staker to allow or deny.
     /// @param _allowed The allowed state.
     function setAllowedStaker(address _staker, bool _allowed) public {
         allowlist[msg.sender][_staker] = _allowed;
         emit BeneficiaryAllowlistUpdated(msg.sender, _staker, _allowed);
+
+        if (!_allowed) {
+            StakedData storage stakingData = stakingData[_staker];
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol]: same here with the naming

File: packages/contracts-bedrock/foundry.toml [hunk 3]
```diff
@@ -74,7 +74,7 @@ fs_permissions = [
 
 # 5159 error code is selfdestruct error code
 ignored_error_codes = ["transient-storage", "code-size", "init-code-size", "too-many-warnings", 5159]
-deny_warnings = true
+deny_warnings = false
```
Comment (0xOneTony) [inline packages/contracts-bedrock/foundry.toml]: why do we change this?

File: packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol [hunk 4]
```diff
@@ -149,102 +180,111 @@ contract PolicyEngineStaking {
...
+        StakedData storage stakingData = stakingData[msg.sender];
+        address currentBeneficiary = stakingData.beneficiary;
+
+        if (currentBeneficiary != _beneficiary) {
+            if (currentBeneficiary != address(0)) {
+                _decreasePeData(currentBeneficiary, stakingData.stakedAmount);
+                emit BeneficiaryRemoved(msg.sender, currentBeneficiary);
+            }
+            stakingData.beneficiary = _beneficiary;
+            emit BeneficiarySet(msg.sender, _beneficiary);
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol:206]: feels like we are also removing some helpful inline comments

**Comment by 0xOneTony** [comment]
will probably need to also run `just pr`

**Review by 0xOneTony — CHANGES_REQUESTED** [review]
We should also downgrade pragma version to 0.8.25 as rev share, requested by Alberto

File: packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol (hunk 1 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol]: you are right [refactor; vars naming](https://github.com/defi-wonderland/optimism/commit/2fc84d9e6b)

File: packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol (hunk 2 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol]: [refactor; vars naming](https://github.com/defi-wonderland/optimism/commit/2fc84d9e6b)

File: packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol (hunk 4 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol:206]: [chore: improve comments](https://github.com/defi-wonderland/optimism/commit/fb1457f716)

**Comment by 0xChin** [comment]
@0xOneTony pragma downgrade handled here [chore downgrade pragma](https://github.com/defi-wonderland/optimism/commit/d485fa5da8)

File: packages/contracts-bedrock/foundry.toml (hunk 3 above)
Comment (0xChin) [reply packages/contracts-bedrock/foundry.toml]: my bad [fix: wrong foundry.toml](https://github.com/defi-wonderland/optimism/commit/3b5dd0fd15)

File: packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol (hunk 2 above)
Comment (0xOneTony) [reply packages/contracts-bedrock/src/periphery/staking/PolicyEngineStaking.sol]: is not actually fixed

**Review by 0xOneTony — APPROVED** [review]
lgtm!

**Review by ashitakah — APPROVED** [review]
